src/type.ml

```diff
@@ -127,10 +124,10 @@ and check_for id r1 r2 body (ctx, d) =
   | _ -> raise (TypeError range_error)
 
 and check_for_impl id r1 r2 body u (ctx, delta) =
-  check_expr r1 (ctx, delta) |> fun (r1_type, (c1, d1)) ->
```

Comment:
Are you sure this is correct? The definition for `check_for` has different logic.